src/ast/ast_validations.c

```diff
@@ -538,665 +702,748 @@ static AST_Validation _ValidateInlinedProperties(const cypher_astnode_t *props)
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+	}
+
+	return true;
+}
+
+// validate an UNWIND clause
+static bool _Validate_UNWIND_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+
+		// introduce alias to bound vars
+		const cypher_astnode_t *alias = cypher_ast_unwind_get_alias(n);
+		const char *identifier = cypher_ast_identifier_get_name(alias);
+		raxInsert(vctx->defined_identifiers, (unsigned char *)identifier, strlen(identifier), NULL, NULL);
+		return true;
+	}
+
+	return true;
+}
+
+// validate a RETURN clause
+static bool _Validate_RETURN_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+		uint num_return_projections = cypher_ast_return_nprojections(n);
+
+		// introduce bound vars
+		for(uint i = 0; i < num_return_projections; i ++) {
+			const cypher_astnode_t *child = cypher_ast_return_get_projection(n, i);
+			const cypher_astnode_t *alias_node = cypher_ast_projection_get_alias(child);
+			if(alias_node == NULL) {
+				continue;
+			}
+			const char *alias = cypher_ast_identifier_get_name(alias_node);
+			raxInsert(vctx->defined_identifiers, (unsigned char *)alias, strlen(alias), NULL, NULL);
+		}
+		return true;
+	}
+
+	if(cypher_ast_return_has_include_existing(n)) {
+		return true;
+	}
+
+	rax           *rax          = raxNew();
+	const char   **columns      = AST_BuildReturnColumnNames(n);
+	uint           column_count = array_len(columns);
+
+	for (uint i = 0; i < column_count; i++) {
+		// column with same name is invalid
+		if(raxTryInsert(rax, (unsigned char *)columns[i], strlen(columns[i]), NULL, NULL) == 0) {
+			ErrorCtx_SetError("Error: Multiple result columns with the same name are not supported.");
+			vctx->valid = AST_INVALID;
+			break;
+		}
+	}
+	
+	raxFree(rax);
+	array_free(columns);
+
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	vctx->valid = _Validate_LIMIT_SKIP_Modifiers(cypher_ast_return_get_limit(n), cypher_ast_return_get_skip(n));
+	
+	return true;
+}
+
+// validate a MATCH clause
+static bool _Validate_MATCH_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
```

Comment:
this type of check, which I assume repeats itself in multiple places should be performed once by the visitor logic, a single place which determines rather or not we need to continue on validating